pyrules/conditions.py

```diff
@@ -0,0 +1,238 @@
+                    bit, expression))
+        for i, bit in enumerate(bits[1:]):
+            try:
+                cur = getattr(cur, bit)
+                continue
+            except AttributeError:
+                try:
+                    cur = cur[bit]
+                    continue
+                except (TypeError, KeyError):
+                    if i == last and bit in self.handlers:
+                        return getattr(self, 'handle_' + bit)(cur, arg)
+                    raise ExpressionError(
+                        'Can\'t resolve "{}" bit in expression: "{}"'.format(
+                            bit, expression))
+        else:
+            return cur == arg
+            
+    handle_gt = operator.gt
+    handle_lt = operator.lt
+    handle_gte = operator.ge
+    handle_lte = operator.le
+    handle_eq = operator.eq
+    handle_neq = operator.ne
+    handle_contains = lambda self, val, arg: arg in val
+    handle_icontains = lambda self, val, arg: arg.lower() in val.lower()
+    handle_bool = lambda self, val, arg: bool(val) == arg
+
+            
+#
+# BNF grammar for parser:
+#
+# number    :: '1' .. '9'
+# negated   :: '~' condition
+# sub_cond  :: '(' condition ')'
+# op        :: '&' | '|' | '^'
+# op_cond   :: condition op condition
+# condition :: number | negated | op_cond | sub_cond
+
+# Trying to convert BNF directly to pyparsing code doesn't work very well -
+# I ended up running into infinite recursion. However, it has good support
+# for math/logic ops parsing with infixNotation helper.        
+
+    
+# parse action maker
+def makeLRlike(numterms=None):
+    """
+    Pyparsing outputs flat lists of token. This helper function converts them
+    to LR-like structures (basically, nested lists) that we can process later.
+
+    Borrowed from http://stackoverflow.com/a/4589920
+    """
+    if numterms is None:
+        # None operator can only by binary op
+        initlen = 2
+        incr = 1
+    else:
+        initlen = {0:1,1:2,2:3,3:5}[numterms]
+        incr = {0:1,1:1,2:2,3:4}[numterms]
+
+    # define parse action for this number of terms,
+    # to convert flat list of tokens into nested list
+    def pa(s,l,t):
+        t = t[0]
+        if len(t) > initlen:
+            ret = ParseResults(t[:initlen])
+            i = initlen
+            while i < len(t):
+                ret = ParseResults([ret] + t[i:i+incr])
+                i += incr
+            return ParseResults([ret])
+    return pa
+
+boolOperand = Word(nums)
+
+# define expression, based on expression operand and
+# list of operations in precedence order
+boolExpr = infixNotation(
+    boolOperand,
+    [
+        ("~", 1, opAssoc.RIGHT, makeLRlike()),
+        ("&", 2, opAssoc.LEFT, makeLRlike(2)),
+        ("|",  2, opAssoc.LEFT, makeLRlike(2)),
+        ("^",  2, opAssoc.LEFT, makeLRlike(2))
+    ])
+
+
+class LogicEvaluator(object):
+    LOGIC_OPS = {'&': operator.and_, '|': operator.or_, '^': operator.xor}
```

Comment:
I very much appreciate the depth of thought you have given this. I had a much simpler approach in mind, but this is certainly the right thing to do. Kudos for taking that path.

Please add a comment on what is the input to this class and what it does. I guess the gist of it is _it parses input expressions into c_expressions`, but what is a logic context, how does it use infixNotation and what does makeLRlike do? No need to write a book about it, just give an example of what goes in what comes out and maybe a few intermediate results in between. Helps the guy who comes after you (which might be yourself of course;-)